contributors/design-proposals/storage/postpone-pvc-deletion-if-used-in-a-pod.md

```diff
@@ -0,0 +1,51 @@
+# Postpone Deletion of a Persistent Volume Claim in case It Is Used by a Pod
+
+Status: Proposal
+
+Version: GA
+
+Implementation Owner: @pospispa
+
+## Motivation
+
+User can delete a Persistent Volume Claim (PVC) that is being used by a pod. This may have negative impact on the pod.
+
+For more details see issue https://github.com/kubernetes/kubernetes/issues/45143
+
+## Proposal
+
+Postpone the PVC deletion until the PVC is not used by any pod.
+
+## User Experience
+
+### Use Cases
+
+1. User deletes a PVC that is being used by a pod. This may have negative impact on the pod. As a user, I want that any PVC deletion does not have any negative impact on any pod.
+
+## Implementation
+
+### PVC Create
+When a PVC is created finalizer is added into its metadata.
+
+### Finalizer
+Finalizer can list pods and watch for pods. It caches status of pods that use a PVC. It also watches for PVC deletes.
+
+Finalizer reacts to PVC delete in one of the below ways:
+- In case the PVC is not used by a pod: removes the finalizer information from the PVC metadata.
+- In case the PVC is used by a pod: sets condition Terminating to true for the PVC.
+
+### Scheduler
+Scheduler will check if a pod uses a PVC and if any of the PVCs has condition Terminating set to true. In case this is true the scheduling of the pod will end up with an error: "PVC (%pvcName) is in Terminating state".
```

Comment:
Can you just check if the PVC has the deletion timestamp set?

Reply:
anything that reacts to absence of a PV or PVC (scheduler, kubelet) should also react to the presence of a deletionTimestamp on a PV or PVC

Reply:
Let me firstly think loudly about the `deletionTimestamp`. The [description of the `deletionTimestamp`](https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#metadata) states: "The resource will be deleted (no longer visible from resource lists, and not reachable by name) after the time in this field." So in case the `deletionTimestamp` is set for a PVC the PVC must be deleted after the time in the `deletionTimestamp`. However, in case the PVC is used by a pod it's not guaranteed that the pod won't be using the PVC after the time in the `deletionTimestamp`. That's why I prefer not to set the `deletionTimestamp` in PVCs, but use the condition Terminating for storing the information that the PVC is going to be deleted.

On the other hand, scheduler must be able to decide whether PVC is going to be deleted or not in order to fail scheduling of a pod that uses the PVC that is going to be deleted.
So it will be better if the condition that PVC is going to be deleted will be: `deletionTimestampIsSet(PVC) || terminatingConditionIsTrue(PVC)`.

Reply:
An object with a deletionTimestamp and non-empty finalizers blocks until finalizers are removed. The description of deletionTimestamp should be updated.

Reply:
I assume that after `kubectl delete pvc pvc-name` the PVC may or may not have the `deletionTimestamp` set. So it's not possible to rely solely on `deletionTimestamp` for finding out whether a PVC is being deleted or not. Am I right?

Reply:
> I assume that after kubectl delete pvc pvc-name the PVC may or may not have the deletionTimestamp set

Not sure what you mean. An effective DELETE API call either sets the deletionTimestamp or completely removes the object from etcd.

Reply:
So in case of a PVC with a finalizer set the DELETE API call will always set the `deletionTimestamp` as the deletion has to be postponed because of the finalizer.
Thus, there can't be a PVC that is being deleted with `deletionTimestamp == nil`.
That said, there is no need for using condition Terminating for marking a PVC that is being deleted.